### PR TITLE
fix(KEN-674): kendex remove disarms the repository before its files go

### PR DESCRIPTION
**Pull request by bmethod**
## Summary

- `kendex remove` runs a leaving package's declared uninstaller while its scripts are still on disk, so removing `growth-guards` disarms the repository instead of leaving hook shims that exec files which are gone. A nonzero uninstaller exit stops the plan with the package still installed.
- One CLI executor owns that ordering: every `EngineReport` goes through `engine_common::apply_report` (uninstall, then apply), so `remove`, `apply`, `refresh`, `marketplace unsubscribe --remove-packages`, source removal and fork rename all disarm. A bare `Plan` with no report behind it drops no package and is executed on its own.
- `kendex check` names the leftover once it exists: a hook file carrying the package's marker in a repository where no project under any skills root holds the package is reported as drift, naming the files to remove. It runs none of the package's scripts.

## Completed Issues

- Closes KEN-674 - kendex remove leaves armed git hooks behind, and the repo cannot commit

## Created Issues

- KEN-709 - Desktop removal leaves a package's armed git hooks behind — Project: CLI & Distribution (blocked by KEN-674, KEN-688)

## Test Plan

- `crates/core/tests/repo_effects_leaving.rs` — the engine reports what leaves a plan, read off the installed tree, including a copy-delivered package
- `crates/cli/tests/install_ux/guarding.rs` — removal disarms first; a failing uninstaller keeps the package installed; applying a manifest without the package disarms first; removal from a plain directory runs no uninstaller
- `crates/cli/tests/guard_hooks/arming.rs` — `check` names the shims a removed package left behind; a neighbouring project carrying the package is not a leftover; a redirected `core.hooksPath` is not reported
- Every new behavior was mutation-checked: each guard's mutant leaves the suite red
- `tools/guard` exits 0

### CODE REVIEW

File: crates/core/src/engine/repo_effects.rs [hunk 1]
```diff
@@ -140,3 +136,94 @@ fn blocked(drift: &[DriftRow], name: &str) -> bool {
...
+        .entries
+        .values()
+        .filter(|entry| entry.kind == ItemKind::Skill && entry.name == name)
+    {
+        for dir in read_dirs(env, scope, entry.harness, ItemKind::Skill) {
+            candidates.push(dir.join(&canonical));
+        }
+    }
+    candidates.into_iter().find_map(|root| {
+        let text = crate::fs::read_if_exists(&root.join("SKILL.md")).ok()??;
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/core/src/engine/repo_effects.rs]: Fixed in 904d00df

File: crates/cli/src/commands/repo_effects.rs [hunk 2]
```diff
@@ -152,3 +145,93 @@ pub fn apply(scope: &Scope, declared: &DeclaredEffects) -> CliResult {
...
+                "{name}: declares no uninstaller — what it changed about this repository stays{}",
+                match &declared.effects.removal {
+                    Some(removal) => format!("; to undo: {}", shown(removal)),
+                    None => String::new(),
+                }
+            ));
+            continue;
+        };
+        say(&format!("{name}: running {}", shown(uninstaller)));
+        let report = relay(scope, declared, uninstaller)?;
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/cli/src/commands/repo_effects.rs]: Declined: the premise is that running the uninstaller needs consent of its own. Removing the package is that consent, and the script is the package's own, from the tree the person installed and can read — the trust model this repo already applies to a rendered package (KEN-672). The separate yes at install exists because arming hooks changes the repository for everyone who commits there; disarming restores what they had, which is why `--allow-repo-effects` gates the install and nothing gates the undo.

Where the effect was never activated there is also nothing to execute against: the installer's `--uninstall` finds no marked line and answers that it removed nothing.

A removal-time opt-in would reintroduce this issue's bug on the path that cannot answer one: a non-interactive `kendex remove` would take the scripts and leave the shims, and every commit in that repository would fail closed. That is the state KEN-674 exists to prevent.

File: crates/core/src/engine/repo_effects.rs [hunk 3]
```diff
@@ -140,3 +136,94 @@ fn blocked(drift: &[DriftRow], name: &str) -> bool {
...
+        Some((root, text))
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/core/src/engine/repo_effects.rs]: Fixed in 904d00df

File: crates/core/src/guard/resolve.rs [hunk 4]
```diff
@@ -85,6 +85,51 @@ impl Installed {
...
+            .iter()
+            .any(|root| carries(root))
+        {
+            return true;
+        }
+        let mut trees = vec![repo.worktree.clone()];
+        if let Some(main) = main.filter(|main| *main != repo.worktree) {
+            trees.push(main);
+        }
+        trees
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/core/src/guard/resolve.rs]: Fixed in 0f5d2f8a

File: crates/core/src/guard/resolve.rs [hunk 5]
```diff
@@ -85,6 +85,51 @@ impl Installed {
...
+        };
+        if search_roots_with(repo, main.clone())
+            .iter()
+            .any(|root| carries(root))
+        {
+            return true;
+        }
+        let mut trees = vec![repo.worktree.clone()];
+        if let Some(main) = main.filter(|main| *main != repo.worktree) {
+            trees.push(main);
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/core/src/guard/resolve.rs]: Fixed in 0f5d2f8a

File: crates/core/src/discover.rs [hunk 6]
```diff
@@ -70,6 +70,40 @@ pub fn discover_projects(root: &Path) -> Result<Vec<PathBuf>> {
...
+/// because a copy at depth six is a copy. `found` is asked before the
+/// descent, so the first hit ends the walk.
+pub fn any_dir(root: &Path, found: &mut dyn FnMut(&Path) -> bool) -> bool {
+    if found(root) {
+        return true;
+    }
+    let Ok(entries) = fs::read_dir(root) else {
+        return false;
+    };
+    for entry in entries.flatten() {
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/core/src/discover.rs]: Fixed in 0f5d2f8a

File: crates/cli/src/commands/engine_common.rs [hunk 7]
```diff
@@ -223,6 +223,21 @@ pub fn confirm_and_apply(
...
+/// A plan can take a package away whatever the verb — `remove`, a manifest
+/// edited by hand and applied, a sweep, an unsubscribe that drops its
+/// packages — and the package's declared uninstaller has to run while the
+/// scripts it names are still on disk. Executing `report.plan` directly
+/// skips that, so no verb does: every report goes through here, and only a
+/// bare `Plan` with no report behind it, which by construction drops no
+/// package, is executed on its own.
+pub fn apply_report(env: &Env, report: &EngineReport) -> Result<usize, Box<dyn std::error::Error>> {
+    super::repo_effects::undo(&report.plan.scope, report)?;
     Ok(kendex_core::apply::execute(env, &report.plan, None)?.applied)
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/cli/src/commands/engine_common.rs:241]: Declined: both states this describes are two kendex runs racing on one machine — another apply owning the scope, or a plan going stale while a confirmation prompt sits open. This repository declines that class rather than defending against it (kendex#1687), and the ordering here is the deliberate one: the uninstaller has to run while the package's scripts are still on disk, which is before the plan executes, not inside it.

The failure it would leave is also the recoverable direction. A disarmed repository whose package is still installed is what `kendex check` reports as not armed, with `kendex guard install` as the remedy. The order this issue exists to prevent is the other one: files gone, shims still armed, every commit failing closed and no automated way back.

File: crates/cli/src/commands/repo_effects.rs [hunk 8]
```diff
@@ -152,3 +145,93 @@ pub fn apply(scope: &Scope, declared: &DeclaredEffects) -> CliResult {
...
+    };
+    let leaving = &report.repo_effects_leaving;
+    // Asked once, and only when something leaving would need the answer:
+    // the probe is git processes, and a package that writes nowhere near
+    // `.git` never asks.
+    let git_here = leaving
+        .iter()
+        .any(|declared| disclose::effects_touch_git(&declared.effects))
+        && disclose::git_dir(root)?.is_some();
+    for declared in leaving {
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/cli/src/commands/repo_effects.rs:182]: Declined: the premise is that kendex should keep the shims when another scope still has the package. growth-guards decided the opposite, in the script this PR calls, and says so at `skills/growth-guards/scripts/install-git-hooks`:

> Uninstall disarms the repository, and asks nothing about who else is here. It used to hunt for a surviving install in another work tree or another nested project and keep the shims for it. Five rounds of review found five holes in that hunt, the last of which — a record of armed projects that only ever grew — left a repository two projects could never disarm at all. Every hole was the same mistake: guessing on someone else's behalf. So consent is repository-level, the level the hooks themselves are at.

The hooks are one directory shared by every work tree and nested project, so arming and disarming are repository-level acts, whoever runs them. Removing the package is a request to undo what it did to this repository, and the package's own uninstaller answers it at that level. A sibling scope that still wants the gate runs `kendex guard install`, which is one command and cannot be wrong about who else is present.

Tracking ownership per scope is the mechanism that was tried and withdrawn; reintroducing it in kendex would rebuild it a level further from the hooks than the package that owns them.

File: crates/core/src/discover.rs [hunk 9]
```diff
@@ -70,6 +70,40 @@ pub fn discover_projects(root: &Path) -> Result<Vec<PathBuf>> {
...
+        let path = entry.path();
+        let Some(name) = path.file_name().and_then(|n| n.to_str()) else {
+            continue;
+        };
+        if name.starts_with('.') || SKIP_DIRS.contains(&name) {
+            continue;
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/core/src/discover.rs:138]: Declined: the pruning is deliberate and the case it hides is one the walk cannot serve. `discover::any_dir` skips hidden directories, the build and dependency trees in `SKIP_DIRS`, and symlinked directories, because a walk that entered `node_modules`, `target` and every dotted directory of a large repository would run at every session start, and this search already sits on `kendex check`, which the drift hook runs then. A vendored copy of some other project's skills tree is also not this repository's install.

The state you describe — a kendex project whose only copy of the package lives under a pruned directory, holding hooks armed from there — yields a wrong advisory, not a broken commit: `check` says to delete files that still work, and the commit chain keeps running. Following the advice is recoverable with `kendex guard install`. That is the class this repository declines rather than defends (kendex#1687), and the hole is in a mechanism that came from review rounds on this PR, not from shipped behavior.

The doc half of your point was right and is corrected: the `Installed::anywhere` rustdoc claimed the search reaches every directory of every tree, which overstated what the walk does; it now says what `discover::any_dir` actually visits.

Reading the helper's baked `installed_scripts` path as the ownership test is a better test than any walk, and it is the right shape for a future change — it answers "is the script these hooks exec still there" directly instead of inferring it. It is not this issue's Done-when, which is that removal runs the uninstaller and `check` names shims a removed package left behind.

File: crates/core/src/discover.rs [hunk 10]
```diff
@@ -70,6 +70,94 @@ pub fn discover_projects(root: &Path) -> Result<Vec<PathBuf>> {
...
+                continue;
+            }
+        };
+        let path = entry.path();
+        let Some(name) = path.file_name().and_then(|n| n.to_str()) else {
+            continue;
+        };
+        if name.starts_with('.') || SKIP_DIRS.contains(&name) {
+            continue;
+        }
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/core/src/discover.rs:139]: Declined: the pruning is deliberate and its cost is a wrong advisory, not a broken repository. `discover::any_dir` skips hidden directories, the build and dependency trees in `SKIP_DIRS`, and symlinked directories because this search runs inside `kendex check`, which the drift hook runs at every session start; a walk that descended into `node_modules`, `target` and every dotted directory of a large repository would be paid there every time.

In the layout you describe the commit chain keeps working — the helper execs its recorded path and the gate runs — while `check` names files it should not. Someone who follows that advice re-arms with one `kendex guard install`. The same reasoning was given on the earlier thread raising this against `.git/hooks` ownership, and the code has not changed since.

The rustdoc that claimed the search reaches every directory of every tree was the part worth fixing, and it is corrected in b04cf9d9: it now says what the walk actually visits, so the negative is not presented as more than it is.

[the remark this answers is not in the document]
Comment (bmethod) [reply crates/core/src/discover.rs:139]: Declined: a kendex project whose directory name is not valid UTF-8, holding the only copy of a package whose hooks are armed, is an input nothing in this product emits. Package and project names are UTF-8 by construction, and the path is one a person would have to build by hand for this to matter.

This repository declines that class rather than defending against it (kendex#1687). The skip is also not new to this diff: the same filter runs on the project walk `any_dir` was built beside.

File: crates/core/src/guard/resolve.rs [hunk 11]
```diff
@@ -85,6 +85,61 @@ impl Installed {
...
+    ///
+    /// [`present`]: Installed::present
+    /// [`Repo::worktrees`]: super::Repo::worktrees
+    pub fn anywhere(repo: &super::Repo) -> Result<bool> {
+        let mut carries = |dir: &Path| {
+            SKILL_ROOTS
+                .iter()
+                .map(|base| dir.join(base).join(SKILL))
+                .any(|dir| dir.exists() || dir.is_symlink())
+        };
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/core/src/guard/resolve.rs:123]: Declined: a dangling link is a broken installation, and `stranded` is not the thing that diagnoses those. It answers one question — did a package leave its shims behind — and a link still standing where the package was installed is not a package that left.

The broken-install state has its own reporting. Where the lock still declares the package, `kendex check` reports it as drift with `kendex refresh` as the remedy, which restores the tree the link points at. Where the package is being removed, the plan takes the link with it. The gap you name is narrower than the thread reads: the lock no longer declares the package, a dangling link survives at exactly a search root, and the hooks are armed — and its cure, `kendex refresh` or a reinstall, is the same one the drift report already prescribes.

Widening `stranded` to judge whether a copy is usable would make it a second opinion on installation health, which `guard check` and the drift report already give, from closer to the evidence.

File: crates/cli/src/commands/repo_effects.rs [hunk 12]
```diff
@@ -152,3 +145,93 @@ pub fn apply(scope: &Scope, declared: &DeclaredEffects) -> CliResult {
...
+        let name = shown(&declared.name);
+        if disclose::effects_touch_git(&declared.effects) && !git_here {
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/cli/src/commands/repo_effects.rs]: Declined: the premise is that the stand-down could keep the non-Git half of the cleanup. There is no half to keep. A package declares one `uninstaller:` — a single command — so the choice at removal is to run that command or not; there is no per-effect granularity to preserve, and running it outside a work tree is what the stand-down exists to avoid, because that is exactly what broke removal in a plain directory earlier on this PR (the installer exits 2 there, and the error kept the package installed forever).

No shipped package has the shape either: `growth-guards` is the one package declaring repo effects, and every write it declares is under `.git`. A package mixing `.git` writes with other repository-relative writes, installed in a work tree that later stops being one, is a layout nothing produces today.

If such a package appears, the honest fix is in the declaration format — effects that can be undone separately need separate uninstallers — not a condition here guessing which parts of one command still apply.

File: crates/core/src/engine/repo_effects.rs [hunk 13]
```diff
@@ -140,3 +137,156 @@ fn blocked(drift: &[DriftRow], name: &str) -> bool {
...
+    if !matches!(scope, Scope::Project { .. }) {
+        return Ok(Vec::new());
+    }
+    let staying = skill_names(after);
+    let mut found: BTreeMap<String, DeclaredEffects> = BTreeMap::new();
+    for name in skill_names(before).difference(&staying) {
+        let Some((root, text)) = installed_tree(env, scope, before, name)? else {
+            continue;
+        };
+        let Some(effects) = declared(&text) else {
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/core/src/engine/repo_effects.rs]: Declined: kendex runs a package's rendered tree as the person installed it, edits included — that is this repository's trust model (KEN-672), not an oversight here. The same objection was raised earlier on this PR against running the uninstaller at all, and the answer is the same: the script comes from the tree the person put in their own repository and can read, and removal is the request to undo what it did.

Proving kendex wrote the bytes would mean the removal path second-guesses an installation that `engine/removal.rs` deliberately accepts as edited. It would also fail closed in the wrong direction: an edited but working package would keep its shims armed with its scripts deleted, which is the state this issue exists to prevent.

File: crates/cli/src/commands/engine_common.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/cli/src/commands/engine_common.rs:241]: Declined: already answered on this PR, at the thread on the same lines raising the scope lock and plan staleness. Nothing in that code has changed since.

Short version: both states are two kendex runs racing on one machine, a class this repository declines rather than defends (kendex#1687). The ordering is deliberate — the uninstaller has to run while the package's scripts are still on disk, which is before the plan executes — and the failure it could leave is the recoverable one: a disarmed repository whose package is still installed is what `kendex check` reports as not armed, with `kendex guard install` as the cure.

File: crates/core/src/engine/repo_effects.rs [hunk 14]
```diff
@@ -140,3 +137,156 @@ fn blocked(drift: &[DriftRow], name: &str) -> bool {
...
+        .entries
+        .values()
+        .filter(|entry| entry.kind == ItemKind::Skill && entry.name == name)
+    {
+        for dir in read_dirs(env, scope, entry.harness, ItemKind::Skill) {
+            candidates.push(dir.join(&canonical));
+        }
+    }
+    for root in candidates {
+        if let Some(text) = crate::fs::read_if_exists(&root.join("SKILL.md"))? {
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/core/src/engine/repo_effects.rs]: Fixed in bfa1f341

File: crates/cli/tests/install_ux/guarding.rs [hunk 15]
```diff
@@ -149,6 +149,227 @@ fn disarming_before_removal_leaves_the_repository_committable() {
...
+        }
+        if dropping && line.starts_with('[') {
+            dropping = false;
+        }
+        if !dropping {
+            kept.push_str(line);
+            kept.push('\n');
+        }
+    }
+    fs::write(world.at("kendex.toml"), kept).unwrap();
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/cli/tests/install_ux/guarding.rs]: Fixed in bfa1f341

File: crates/core/src/engine/repo_effects.rs [hunk 16]
```diff
@@ -140,3 +137,156 @@ fn blocked(drift: &[DriftRow], name: &str) -> bool {
...
+        return Ok(Vec::new());
+    }
+    let staying = skill_names(after);
+    let mut found: BTreeMap<String, DeclaredEffects> = BTreeMap::new();
+    for name in skill_names(before).difference(&staying) {
+        let Some((root, text)) = installed_tree(env, scope, before, name)? else {
+            continue;
+        };
+        let Some(effects) = declared(&text) else {
+            continue;
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/core/src/engine/repo_effects.rs]: Fixed in 36f84819

File: crates/core/src/discover.rs [hunk 17]
```diff
@@ -70,6 +70,94 @@ pub fn discover_projects(root: &Path) -> Result<Vec<PathBuf>> {
...
+        let entry = match entry {
+            Ok(entry) => entry,
+            Err(error) => {
+                unreadable.get_or_insert_with(|| CoreError::io(dir, error));
+                continue;
+            }
+        };
+        let path = entry.path();
+        let Some(name) = path.file_name().and_then(|n| n.to_str()) else {
+            continue;
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/core/src/discover.rs:135]: Declined: already answered on this PR, at the thread raising the same skip against `any_dir`. Nothing in that code has changed since.

Short version: a kendex project whose directory name is not valid UTF-8, holding the only copy of a package whose hooks are armed, is an input nothing in this product emits — package and project names are UTF-8 by construction, and the path has to be built by hand for this to matter. This repository declines that class rather than defending against it (kendex#1687). The filter is also not new here: the same one runs on the project walk `any_dir` was built beside.

File: crates/core/src/engine/repo_effects.rs [hunk 18]
```diff
@@ -135,3 +137,230 @@ fn blocked(drift: &[DriftRow], name: &str) -> bool {
...
+    let staying = skill_names(after);
+    let mut found: BTreeMap<String, DeclaredEffects> = BTreeMap::new();
+    for name in skill_names(before).difference(&staying) {
+        let Some(installed) = installed_tree(env, scope, before, name)? else {
+            continue;
+        };
+        let effects = match declaration(&installed.text) {
+            Declaration::Effects(effects) => effects,
+            // A package that declares nothing has nothing to undo.
+            Declaration::Absent => continue,
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/core/src/engine/repo_effects.rs:181]: Declined: this asks the removal path to stop believing the installed tree, which is the opposite of the answer already given on this PR to the thread asking it to prove kendex wrote those bytes. kendex runs a package as the person installed it, edits included (KEN-672), and removal reads the same copy for the same reason: it is the copy whose scripts are on disk and whose installer armed the repository.

Deriving the departing effects from the source record instead would make the two halves disagree — the disclosure at install time reads the tree, the undo at removal time would read something else — and it would run an uninstaller the installed package no longer declares, which is its own way of surprising someone.

The state you describe also needs a person to edit the `repo-effects` block out of an installed package while its hooks are armed, and then remove it with `--discard-edits`. The three-state parser it turns on was itself added by an earlier round of this review; a hole in that mechanism is declined here rather than fixed by widening it again.

If the authoritative declaration for removal should be the source rather than the install, that is a product decision about what an edited package means, and it belongs in its own issue with its own Done-when — not in a fix whose scope is that removal runs the uninstaller while the scripts are still there.

File: crates/core/src/guard/mod.rs [hunk 19]
```diff
@@ -213,29 +217,82 @@ fn installer(dir: &Path, args: &[&str]) -> Result<GuardReport> {
...
+/// Empty where `core.hooksPath` is set: git reads no hook here, so nothing
+/// fails, and what a redirected directory means is a grammar this module
+/// does not have. The execute bit is not consulted: a leftover git happens
+/// to skip is still a leftover.
+pub fn stranded(repo: &Repo) -> Result<Vec<PathBuf>> {
+    let hooks = repo.default_hooks_dir();
+    let mut files = Vec::new();
+    for lane in LANES {
+        let path = hooks.join(lane);
+        if crate::fs::read_if_exists(&path)?.is_some_and(|text| text.contains(MARKER)) {
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/core/src/guard/mod.rs]: Fixed in 474c8fd9

File: crates/core/src/repo_effects/declaration.rs [hunk 20]
```diff
@@ -24,12 +24,8 @@ pub struct RepoEffects {
...
     /// ends with what the reader should run themselves.
     pub installer: Option<String>,
-    /// The script that undoes the effect.
-    ///
-    /// Declared, not yet run: nothing in kendex executes it, and `remove`
-    /// takes the package's files away with the effect still applied. The
-    /// disclosure names it so a person can run it themselves, which is the
-    /// whole of what it does today. KEN-674 carries wiring it into removal.
+    /// The script that undoes the effect. A plan that takes the package
+    /// out of a scope runs it first, while the file is still there.
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/core/src/repo_effects/declaration.rs]: Fixed in 474c8fd9

File: crates/core/src/repo_effects/declaration.rs [hunk 21]
```diff
@@ -44,25 +40,80 @@ pub struct RepoEffects {
...
+    // Frontmatter that will not parse is a declaration kendex could not
+    // read, never one that is not there: `parse_tolerant` fails the whole
+    // block for any multi-line entry whose YAML is broken, and a
+    // `repo-effects` block is always multi-line — so a missing key here is
+    // a key that was never looked for.
+    let Ok(parsed) = crate::frontmatter::parse_tolerant(yaml) else {
+        return Declaration::Unreadable;
+    };
+    let Some(value) = parsed.map.get(KEY) else {
+        return Declaration::Absent;
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/core/src/repo_effects/declaration.rs]: Fixed in 9d9419ab

File: crates/core/src/engine/repo_effects.rs [hunk 22]
```diff
@@ -135,3 +137,230 @@ fn blocked(drift: &[DriftRow], name: &str) -> bool {
...
+) -> Result<Option<Installed>> {
+    let canonical = crate::harness::canonical_name(name);
+    let mut candidates = vec![skill_canonical(env, scope, name)];
+    for entry in before
+        .entries
+        .values()
+        .filter(|entry| entry.kind == ItemKind::Skill && entry.name == name)
+    {
+        for dir in read_dirs(env, scope, entry.harness, ItemKind::Skill) {
+            candidates.push(dir.join(&canonical));
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/core/src/engine/repo_effects.rs]: Fixed in 9d9419ab

File: crates/cli/src/commands/repo_effects.rs [hunk 23]
```diff
@@ -118,3 +121,96 @@ fn relay(report: &kendex_core::guard::GuardReport) {
...
+                "{name}: declares no uninstaller — what it changed about this repository stays{}",
+                match &declared.effects.removal {
+                    Some(removal) => format!("; to undo: {}", shown(removal)),
+                    None => String::new(),
+                }
+            ));
+            continue;
+        };
+        say(&format!("{name}: running {}", shown(uninstaller)));
+        let report = run(scope, declared, uninstaller)?;
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/cli/src/commands/repo_effects.rs:201]: Declined: already answered on this PR, at the thread raising the same case against `undo`. Nothing in that code has changed since.

Short version: growth-guards decided this, in the script this PR calls. Its uninstaller "disarms the repository, and asks nothing about who else is here" — it used to hunt for a surviving install in another work tree or nested project and keep the shims for it, five rounds of review found five holes in that hunt, and the last left a repository two projects could never disarm at all. Consent for these hooks is repository-level because the hooks are. A sibling scope that still wants the gate runs `kendex guard install`.

`stranded` documenting that every project and worktree shares the directory is the same fact from the diagnose side, and it is why removing from one scope is a repository-level act rather than a per-scope one.

File: crates/cli/src/commands/engine_common.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/cli/src/commands/engine_common.rs:241]: Declined: this failure leaves the recoverable state, and it is not silent. A repository whose hooks were disarmed while the package is still installed is exactly what `kendex check` reports as not armed, naming `kendex guard install` as the remedy — the same report a plain `kendex guard uninstall` produces. The direction this issue exists to prevent is the other one: scripts gone, shims armed, every commit failing closed with nothing to run.

The ordering is deliberate and cannot be reversed without reintroducing that: the uninstaller has to run while the package's scripts are still on disk, which is before the plan trashes them.

Re-arming automatically after a failed apply would mean running the installer on a repository whose state kendex has just failed to establish, without the consent that arming asks for. This repository's answer to a partial failure is to report it and let the person choose, not to guess.

File: crates/core/src/engine/repo_effects.rs [hunk 24]
```diff
@@ -135,3 +137,230 @@ fn blocked(drift: &[DriftRow], name: &str) -> bool {
...
+/// call a package that declares an uninstaller a package that declares
+/// nothing.
+fn installed_tree(
+    env: &Env,
+    scope: &Scope,
+    before: &crate::lock::Lock,
+    name: &str,
+) -> Result<Option<Installed>> {
+    let canonical = crate::harness::canonical_name(name);
+    let mut candidates = vec![skill_canonical(env, scope, name)];
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/core/src/engine/repo_effects.rs:261]: Declined, with the reasoning stated rather than waved at, because the mechanism you name is real.

Reaching it needs three things at once: the package delivered by copy, so the shared tree holds no copy kendex wrote; an unmanaged skill of the same name sitting in that shared tree; and that skill declaring a `repo-effects` block of its own. A tree with the first two and not the third loses only the uninstaller run, which is the case the earlier read-error and disabled-name fixes on this PR already narrowed.

This PR is at 1.97x the diffstat its first review was posted against, which is the point where this repository's rule stops adding code to a change and starts cutting it. A compound edge in a mechanism that itself came from these review rounds is what that rule is for; widening the candidate resolution again is how the previous four rounds each opened the next one.

If the emitted tree should be the authority for resolving a departing package — rather than a candidate list ordered by convention — that is worth doing properly against the lock's own record of what was written, with its own Done-when. It is not something to bolt onto a removal fix at this size.

File: crates/cli/src/commands/check.rs [hunk 25]
```diff
@@ -61,24 +61,60 @@ fn fold_commit_hooks(env: &Env, checked: &mut CheckReport, scopes: &[kendex_core
...
+                        .map(|path| path.display().to_string())
+                        .collect();
+                    (
+                        Class::Drift,
+                        format!(
+                            "{} armed the commit hooks and is installed in no project of this repository, so every commit fails — delete {} or strip the lines marked `{}` from the hooks (a hook the marker says it created can go whole)",
+                            kendex_core::guard::SKILL,
+                            files.join(", "),
+                            kendex_core::guard::MARKER
+                        ),
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/cli/src/commands/check.rs]: Fixed in 7af668b6

File: crates/core/src/engine/repo_effects.rs [hunk 26]
```diff
@@ -135,3 +137,157 @@ fn blocked(drift: &[DriftRow], name: &str) -> bool {
...
+        let rendered = crate::harness::rendered_name(entry.harness, name);
+        for dir in read_dirs(env, scope, entry.harness, ItemKind::Skill) {
+            candidates.push(dir.join(&rendered));
+        }
+    }
+    for root in candidates {
+        for file in DECLARATION_NAMES {
+            let declaration = root.join(file);
+            if let Some(text) = crate::fs::read_if_exists(&declaration)? {
+                return Ok(Some(Installed {
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/core/src/engine/repo_effects.rs:281]: Declined, and your precedent is accurate: `fork::forkable::ambiguous_skill_tree` already calls an installed tree carrying both spellings ambiguous, and `both_names` refuses a catalog that ships both. Neither covers the case you describe — a tree that gained the second file by hand after installation — so the gap is real rather than already-guarded.

What stops it here is what the answer would have to be. Refusing a removal because someone added a file by hand is its own surprise: the person asked for the package to go, and the reply would be that kendex will not take it. Preferring the disabled declaration silently is the other half of the same decision. Which of those is right is a product question about what an installed tree carrying two claims means, and this issue's Done-when does not carry it — it says removal runs the uninstaller while the scripts are still there, and `check` names what a removed package left behind.

The precondition is also compound: the package installed and armed, then switched off, then given a plain `SKILL.md` by hand carrying no `repo-effects` block.

And this PR is at 1.97x the diffstat its first review was posted against, which is where this repository's rule stops adding to a change. Three findings before yours were recorded rather than fixed for the same reason; a fourth that needs a new decision about ambiguity is not the one to break it for.

Reusing `ambiguous_skill_tree` at the removal read is the shape a follow-up should take, once what it should then do is settled.

File: crates/cli/src/commands/engine_common.rs [hunk 27]
```diff
@@ -223,6 +223,21 @@ pub fn confirm_and_apply(
...
+///
+/// A plan can take a package away whatever the verb — `remove`, a manifest
+/// edited by hand and applied, a sweep, an unsubscribe that drops its
+/// packages — and the package's declared uninstaller has to run while the
+/// scripts it names are still on disk. Executing `report.plan` directly
+/// skips that, so no verb does: every report goes through here, and only a
+/// bare `Plan` with no report behind it, which by construction drops no
+/// package, is executed on its own.
+pub fn apply_report(env: &Env, report: &EngineReport) -> Result<usize, Box<dyn std::error::Error>> {
+    super::repo_effects::undo(&report.plan.scope, report)?;
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/cli/src/commands/engine_common.rs:240]: Tracked: KEN-709

The desktop paths are that issue's scope — it names `audit.rs` `remove_item` and `apply_scope`, `unsubscribe.rs` `marketplace_unsubscribe`, the removal-capable report paths in `packages.rs`, and `sources.rs`, and its Done-when is that a desktop removal leaves nothing armed behind or refuses and names the CLI verb, with the app's result saying which. It is blocked on this PR, which creates the `repo_effects_leaving` it consumes, and on KEN-688, which merged today.

File: crates/core/src/discover.rs [hunk 28]
```diff
@@ -70,6 +70,94 @@ pub fn discover_projects(root: &Path) -> Result<Vec<PathBuf>> {
...
+        // reads as not-a-directory and is pruned by the same test — one
+        // stat off the entry rather than two off the path. An entry whose
+        // type could not be read might have been a directory holding what
+        // is being looked for, so it counts as unseen rather than as a
+        // file.
+        match entry.file_type() {
+            Ok(kind) if kind.is_dir() => {
+                if walk_any(&path, found, unreadable) {
+                    return true;
+                }
```
[the remark this answers is not in the document]
Comment (bmethod) [reply crates/core/src/discover.rs:150]: Declined here, and it is the sharpest finding of this round: it is a false negative, where the others were false positives. A copy inside a submodule arms that submodule's hooks, not the outer repository's, so counting it as an owner makes `kendex check` report clean while every commit in the outer repository fails — the diagnosis this issue exists for, silenced.

Two things stop it going in now. The boundary is subtler than it reads: `Repo::worktrees` hands the walk each work tree's own root, and a linked work tree's root carries a `.git` entry of its own, so "skip a directory holding `.git`" written the obvious way skips every root the walk was given. The rule has to be nested-below-the-root, not any-`.git`, and it needs a test with a submodule and a linked worktree side by side to show it kept the second while dropping the first.

And this PR stands at +2119 against the +1076 its first review was posted on. Two per cent of headroom is not where a walk's boundary conditions get rewritten; the last four rounds each came from adjusting this same walk, and each adjustment looked this small.

So it belongs in a follow-up with room for that test, not here. I am recording it in the PR summary rather than filing, per the owner's call on this class — but unlike the pruning and non-UTF-8 cases, this one silences the report rather than over-reporting, which is the direction worth fixing first when someone picks it up.
